README.md
# sofutobanku
Setup utility for a certain Hikari provider in Japan

## Linux server setup

This explains a setup that uses the NetworkManager stack to connect to the
Internet. Distributions like Fedora Linux use this software to control the
network stack.

* External interface
  * Dibbler is the only DHCPv6 client suitable for our use. Make sure to put
    the Auth Server, Shared Secret, and Password in `/etc/sysconfig/sofutobanku`.
    Use the following config file with `<Internet interface>` substituted for
    the right values for your configuration.

`/etc/dibbler/client.conf`:
```
# Dibbler client config for SoftBank Hikari
duid-type duid-ll
inactive-mode
skip-confirm
log-mode short
log-level 7
script "/etc/softubanku/dibbler.sh"
t1 0
t2 0
reconfigure-accept 1

# You can specify downlink interfaces:
#downlink-prefix-ifaces "eth1", "eth2", "wifi0"
# Or set it off to manually configure them elsewhere:
#downlink-prefix-ifaces "none"

iface "<Internet interface>" {
  pd
  option dns-server
  option domain
  option ntp-server
  option vendor-spec
}
```
 
## Normal setup flow

This section discusses the flow needed to fully set up the Internet connection
and have all the information necessary to bring up the SIP connection if
desired.

1. ICMPv6 Router Solicitation
   1. Receive MTU information from router (i.e., 1500 bytes; see section
      2.4.2.1.5 of [FLETS])
1. IPv4 DHCP exchange (NTT SIP network)
   1. Local address for use with SIP
   1. SIP server address
   1. Static route for SIP network
1. IPv6 DHCP exchange (Internet)
   * Request:
      1. Client ID must be of the DUID-LL (Link Layer) type (see section
         2.4.2.1.4 of [FLETS])
         * Format is `00:03:00:01:<6-byte MAC address>`
      1. Request should include Prefix Delegation (PD) (see section 2.4.2.1.2
         of [FLETS])
   * Response:
      1. Vendor-specific information (NTT):
         1. MAC address (option 201)
         1. Hikari denwa telephone number (option 202)
         1. SIP domain (option 204)
         1. Route information (option 210; not needed?)
      1. Identity Assocation for Prefix Delegation (IA-PD)
         * Sends a /56 network
         * Internal LAN address should be set to PD prefix in this format:
           `xxxx:xxxx:xxxx:xx00:1111:1111:1111:1111/64` (*not* on the interface
           it received the delegation from)
1. IPv6 RADIUS exchange (IPv4-in-IPv6 setup)
   * Access-Request (1) packet:
      1. RADIUS Shared Secret and Password is needed
      1. Contains IA-PD prefix as username
         * Format is `xxxx:xxxx:xxxx:xx00:1111:1111:1111:1111`
      1. Must contain Vendor Specific Attributes (VSA)
         * MAC Address (1)
         * Client manufacturer (2)
         * Client software version (3)
         * Client hardware revision (4)
      1. CHAP authentication
         * Uses CHAP-Challenge (60) attribute
         * Password is shared among all clients
   * Access-Accept (2) packet:
      1. Contains Vendor Specific Attributes (VSA)
         * IPv4-in-IPv6 tunnel local IPv4 address (204)
         * IPv4-in-IPv6 tunnel endpoint IPv6 address (207)
      1. Other attributes don't appear to be useful
1. IPv6-in-IPv4 tunnel setup
   1. Use IPv4-in-IPv6 parameters discovered in IPv6 RADIUS exchange
   1. Must NOT have [Tunnel Encapsulation Limit Option][tunnel-encap]
      * Requires NetworkManager 1.12 or newer ([link to bug][nm-bug])

[FLETS]: https://flets.com/pdf/ip-int-flets-3.pdf "IP Network Service Interface for FLETS"
[tunnel-encap]: https://tools.ietf.org/html/rfc2473#page-13 "Tunnel Encapsulation Limit Option"
[nm-bug]: https://bugzilla.gnome.org/show_bug.cgi?id=791846 "NetworkManager bug"


ChangeLog.txt
* Mon Mar 5 2018 Kenny Root <kenny@the-b.org> 1.0.0-1
- First release.


ip_address_tests.cc
#include <sstream>

#include <gtest/gtest.h>

#include "ip_address.h"

TEST(IpAddress, GetBracketed_IPv6HasBrackets) {
  IpAddress addr("2001:db8::9:139");
  ASSERT_EQ("[2001:db8::9:139]", addr.get_bracketed());
}

TEST(IpAddress, GetBracketed_LeavesIPv4Alone) {
  IpAddress addr("127.0.0.1");
  ASSERT_EQ("127.0.0.1", addr.get_bracketed());
}

TEST(IpAddress, IsIPv6_IPv6_Success) {
  IpAddress addr("2001:db8:1000::1");
  ASSERT_TRUE(addr.is_ipv6());
}

TEST(IpAddress, IsIPv6_IPv4_Success) {
  IpAddress addr("127.0.0.1");
  ASSERT_FALSE(addr.is_ipv6());
}

TEST(IpAddress, GetExpanded_IPv6_Success) {
  IpAddress addr("2001:db8:9:100::39");
  ASSERT_EQ("2001:0db8:0009:0100:0000:0000:0000:0039", addr.get_expanded());
}

TEST(IpAddress, GetExpanded_IPv4_Same) {
  IpAddress addr("127.0.0.1");
  ASSERT_EQ("127.0.0.1", addr.get_expanded());
}

TEST(IpAddress, Stream_IPv4) {
  IpAddress addr("127.0.0.1");
  std::stringstream output;
  output << addr;
  ASSERT_EQ("127.0.0.1", output.str());
}

TEST(IpAddress, Stream_IPv6) {
  IpAddress addr("2001:db8::9:139");
  std::stringstream output;
  output << addr;
  ASSERT_EQ("2001:0db8:0000:0000:0000:0000:0009:0139", output.str());
}


radius.h
/*
 * Copyright 2018 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef _SB_CLIENT_RADIUS_H
#define _SB_CLIENT_RADIUS_H

#include "ip_address.h"

int radius_transact(const IpAddress &auth_server_ip,
                    const std::string &shared_secret, const IpAddress &username,
                    const std::string &password, const std::string &mac);

#endif /* _SB_CLIENT_RADIUS_H */


src/main.cc
/*
 * Copyright 2018 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <cstring>
#include <iomanip>
#include <iostream>

#include "argh.h"
#include "radius.h"

void print_usage(std::string executable_name) {
  std::cerr << "S*ftB*nk Hikari RADIUS authenticator" << std::endl;
  std::cerr << "Usage:\t" << executable_name
            << " --ip <local IP> --mac <local MAC> " << std::endl;
  std::cerr
      << "\t\t--auth-server <RADIUS server IP> --shared-secret <RADIUS secret> "
      << std::endl;
  std::cerr << "\t\t--password <RADIUS password>" << std::endl;
}

int main(int argc, char **argv) {
  auto cmdl = argh::parser(argv, argh::parser::PREFER_PARAM_FOR_UNREG_OPTION);

  if (cmdl["help"]) {
    print_usage(cmdl[0]);
    return 1;
  }

  bool need_usage = false;
  if (cmdl("ip").str().empty()) {
    std::cerr << "Missing local IP address" << std::endl;
    need_usage = true;
  }
  if (cmdl("mac").str().empty()) {
    std::cerr << "Missing local MAC address" << std::endl;
    need_usage = true;
  }
  if (cmdl("auth-server").str().empty()) {
    std::cerr << "Missing RADIUS auth server IP address" << std::endl;
    need_usage = true;
  }
  if (cmdl("shared-secret").str().empty()) {
    std::cerr << "Missing RADIUS shared secret" << std::endl;
    need_usage = true;
  }
  if (cmdl("password").str().empty()) {
    std::cerr << "Missing RADIUS password" << std::endl;
    need_usage = true;
  }

  if (need_usage) {
    std::cerr << std::endl;
    print_usage(cmdl[0]);
    return 1;
  }

  return radius_transact(cmdl("auth-server").str(), cmdl("shared-secret").str(),
                         cmdl("ip").str(), cmdl("password").str(),
                         cmdl("mac").str());
}


src/radius.cc
/*
 * Copyright 2018 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cstring>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <memory>
#include <string>

#include <arpa/inet.h>
#include <netdb.h>

#include <nettle/md5.h>
#include <radcli/radcli.h>

#include "config.h"
#include "radius.h"

const std::string kManufacturer = "foxconn";
const std::string kModel = "e-wmta2.3,V5.0.0.1.rc35";
const std::string kHwRev = "hw_rev_2.00";

constexpr int kVendorSoftbank = 22197;

constexpr int kSbBBMac = 1;
constexpr int kSbBBManufacturer = 2;
constexpr int kSbBBModel = 3;
constexpr int kSbBBHWRev = 4;

constexpr int kSbIPv4LocalAddr = 204;
constexpr int kSbIPv4TunnelEndpoint = 207;

int radius_transact(const IpAddress &auth_server_ip,
                    const std::string &shared_secret, const IpAddress &username,
                    const std::string &password, const std::string &mac) {
  unique_rc_handle rh(rc_new());
  if (rh.get() == nullptr) {
    std::cerr << "ERROR: unable create new handle" << std::endl;
    return ERROR_RC;
  }

  if (rc_config_init(rh.get()) == nullptr) {
    std::cerr << "%s: unable to init rc" << std::endl;
    return ERROR_RC;
  }

  if (rc_add_config(rh.get(), "dictionary", "/etc/radcli/dictionary", "config",
                    0) != 0) {
    std::cerr << "ERROR: Unable to set dictionary" << std::endl;
    return ERROR_RC;
  }

  std::string auth_server;
  auth_server += auth_server_ip.get_bracketed();
  auth_server += "::";
  auth_server += shared_secret;
  if (rc_add_config(rh.get(), "authserver", auth_server.c_str(), "config", 0) !=
      0) {
    std::cerr << "ERROR: unable to set authserver" << std::endl;
    return ERROR_RC;
  }

  if (rc_add_config(rh.get(), "radius_retries", "3", "config", 0) != 0) {
    std::cerr << "ERROR: Unable to set radius_retries." << std::endl;
    return ERROR_RC;
  }

  if (rc_add_config(rh.get(), "radius_timeout", "5", "config", 0) != 0) {
    std::cerr << "ERROR: Unable to set radius_timeout." << std::endl;
    return ERROR_RC;
  }

  if (rc_test_config(rh.get(), "config") != 0) {
    std::cerr << "ERROR: config incomplete" << std::endl;
    return ERROR_RC;
  }

  if (rc_read_dictionary(rh.get(), rc_conf_str(rh.get(), "dictionary")) != 0) {
    std::cerr << "ERROR: Failed to initialize radius dictionary" << std::endl;
    return ERROR_RC;
  }

  std::string dictionary_dir(SB_CLIENT_SYSCONFDIR);
  dictionary_dir += "/dictionary.softbank";
  if (rc_read_dictionary(rh.get(), dictionary_dir.c_str()) != 0) {
    std::cerr << "ERROR: cannot read SoftBank dictionary from "
              << dictionary_dir << std::endl;
    return ERROR_RC;
  }

  std::string expanded_user_ip = username.get_expanded();
  unique_VALUE_PAIR send(rc_avpair_new(rh.get(), PW_USER_NAME,
                                       expanded_user_ip.c_str(),
                                       expanded_user_ip.size(), VENDOR_NONE));
  if (send.get() == nullptr) {
    std::cerr << "%s: unable to add username" << std::endl;
    return ERROR_RC;
  }

  // Okay, this is gross, but the API sucks for this.
  VALUE_PAIR *send_raw = send.get();

  std::unique_ptr<char> challenge_bytes(new char[MD5_DIGEST_SIZE]);
  std::ifstream randin("/dev/urandom");
  randin.read(challenge_bytes.get(), MD5_DIGEST_SIZE);
  randin.close();

  std::string challenge;
  challenge.insert(0, challenge_bytes.get(), MD5_DIGEST_SIZE);
  if (rc_avpair_add(rh.get(), &send_raw, PW_CHAP_CHALLENGE, challenge.c_str(),
                    challenge.size(), VENDOR_NONE) == nullptr) {
    std::cerr << "ERROR: cannot add CHAP challenge to packet" << std::endl;
    return ERROR_RC;
  }

  std::string chap_password;
  chap_password += '\x01';
  chap_password += password;
  chap_password += challenge;

  std::unique_ptr<struct md5_ctx> ctx(new struct md5_ctx);
  md5_init(ctx.get());
  md5_update(ctx.get(), chap_password.size(),
             reinterpret_cast<const uint8_t *>(chap_password.c_str()));
  std::unique_ptr<uint8_t> hashed_bytes(new uint8_t[MD5_DIGEST_SIZE]);
  md5_digest(ctx.get(), MD5_DIGEST_SIZE, hashed_bytes.get());

  std::string response;
  response += '\x01';
  response.insert(1, reinterpret_cast<const char *>(hashed_bytes.get()),
                  MD5_DIGEST_SIZE);

  if (rc_avpair_add(rh.get(), &send_raw, PW_CHAP_PASSWORD, response.c_str(),
                    response.size(), VENDOR_NONE) == nullptr) {
    std::cerr << "ERROR: unable to add password" << std::endl;
    return ERROR_RC;
  }

  if (rc_avpair_add(rh.get(), &send_raw, kSbBBMac, mac.c_str(), mac.size(),
                    kVendorSoftbank) == nullptr) {
    std::cerr << "ERROR: unable to set MAC" << std::endl;
    return ERROR_RC;
  }

  std::string manufacturer(kManufacturer);
  if (rc_avpair_add(rh.get(), &send_raw, kSbBBManufacturer,
                    manufacturer.c_str(), manufacturer.size(),
                    kVendorSoftbank) == nullptr) {
    std::cerr << "ERROR: unable to set manufacturer" << std::endl;
    return ERROR_RC;
  }

  std::string model(kModel);
  if (rc_avpair_add(rh.get(), &send_raw, kSbBBModel, model.c_str(),
                    model.size(), kVendorSoftbank) == nullptr) {
    std::cerr << "ERROR: unable to set model" << std::endl;
    return ERROR_RC;
  }

  std::string hwrev(kHwRev);
  if (rc_avpair_add(rh.get(), &send_raw, kSbBBHWRev, hwrev.c_str(),
                    hwrev.size(), kVendorSoftbank) == nullptr) {
    std::cerr << "ERROR: unable to set HW revision" << std::endl;
    return ERROR_RC;
  }

  VALUE_PAIR *received;
  int result =
      rc_aaa(rh.get(), 0, send.get(), &received, nullptr, 0, PW_ACCESS_REQUEST);
  if (result == OK_RC) {
    unique_VALUE_PAIR vp(received);

    VALUE_PAIR *vp_local_address =
        rc_avpair_get(vp.get(), kSbIPv4LocalAddr, kVendorSoftbank);
    if (vp_local_address == nullptr) {
      std::cerr << "Local Address attribute is missing: " << kVendorSoftbank
                << "/" << kSbIPv4LocalAddr << std::endl;
    }

    VALUE_PAIR *vp_tunnel_endpoint =
        rc_avpair_get(vp.get(), kSbIPv4TunnelEndpoint, kVendorSoftbank);
    if (vp_tunnel_endpoint == nullptr) {
      std::cerr << "Tunnel Endpoint attribute is missing: " << kVendorSoftbank
                << "/" << kSbIPv4TunnelEndpoint << std::endl;
    }

    char junk[1];
    char local_address[INET_ADDRSTRLEN];
    char tunnel_endpoint[INET6_ADDRSTRLEN];

    if (rc_avpair_tostr(rh.get(), vp_local_address, junk, sizeof(junk),
                        local_address, sizeof(local_address)) ||
        rc_avpair_tostr(rh.get(), vp_tunnel_endpoint, junk, sizeof(junk),
                        tunnel_endpoint, sizeof(tunnel_endpoint))) {
      std::cerr << "Cannot find Local IPv4 Address or Tunnel Endpoint!"
                << std::endl;
      return ERROR_RC;
    }

    std::cout << local_address << " " << tunnel_endpoint << std::endl;
  } else {
    std::cerr << username << " RADIUS Authentication failure (RC=" << result
              << ")" << std::endl;
  }

  return result;
}


unique_types.h
/*
 * Copyright 2018 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef _SB_CLIENT_UNIQUE_TYPES_H
#define _SB_CLIENT_UNIQUE_TYPES_H

#include <memory>
#include <radcli/radcli.h>

struct rc_handle_Deleter {
  void operator()(rc_handle *p) { rc_destroy(p); }
};

struct VALUE_PAIR_Deleter {
  void operator()(VALUE_PAIR *p) { rc_avpair_free(p); }
};

struct addrinfo_Deleter {
  void operator()(addrinfo *p) { freeaddrinfo(p); }
};

using unique_rc_handle = std::unique_ptr<rc_handle, rc_handle_Deleter>;
using unique_VALUE_PAIR = std::unique_ptr<VALUE_PAIR, VALUE_PAIR_Deleter>;
using unique_addrinfo_ret = std::unique_ptr<addrinfo, addrinfo_Deleter>;

#endif /* _SB_CLIENT_UNIQUE_TYPES_H */


src/ip_address.h
/*
 * Copyright 2018 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef _SB_CLIENT_IP_ADDRESS_H
#define _SB_CLIENT_IP_ADDRESS_H

#include "unique_types.h"
#include <memory>
#include <string>

class IpAddress {
public:
  IpAddress(std::string);
  IpAddress(const char *);

  std::string get_bracketed() const;
  bool is_ipv6() const;
  std::string get_expanded() const;

  friend std::ostream &operator<<(std::ostream &os, const IpAddress &ip);

private:
  unique_addrinfo_ret ai_;

  bool resolve_and_set_address(const char *);
};

#endif /* _SB_CLIENT_IP_ADDRESS_H */


src/ip_address.cc
/*
 * Copyright 2018 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cstring>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>

#include <arpa/inet.h>
#include <netdb.h>
#include <netinet/in.h>

#include "ip_address.h"
#include "unique_types.h"

IpAddress::IpAddress(std::string ip) { resolve_and_set_address(ip.c_str()); }

IpAddress::IpAddress(const char *ip) { resolve_and_set_address(ip); }

bool IpAddress::resolve_and_set_address(const char *ip) {
  std::unique_ptr<addrinfo> hint(new addrinfo);
  std::memset(hint.get(), 0, sizeof(addrinfo));
  hint.get()->ai_family = PF_UNSPEC;
  hint.get()->ai_flags = AI_NUMERICHOST | AI_CANONNAME;

  struct addrinfo *res;
  int ret = getaddrinfo(ip, NULL, hint.get(), &res);
  if (ret) {
    std::cerr << "ERROR: invalid address given (" << gai_strerror(ret)
              << "): " << ip << std::endl;
    return false;
  }

  ai_.reset(res);
  return true;
}

std::string IpAddress::get_bracketed() const {
  std::unique_ptr<char> hbuf(new char[NI_MAXHOST]);
  if (getnameinfo(ai_.get()->ai_addr, ai_.get()->ai_addrlen, hbuf.get(),
                  NI_MAXHOST, nullptr, 0, NI_NUMERICHOST) != 0) {
    return "invalid";
  }

  std::string ip_converted;
  if (ai_.get()->ai_family == AF_INET6) {
    std::string bracketed;
    bracketed += '[';
    bracketed += hbuf.get();
    bracketed += ']';
    return bracketed;
  } else {
    return hbuf.get();
  }
}

std::string IpAddress::get_expanded() const {
  std::stringstream expanded;
  if (!is_ipv6()) {
    return ai_.get()->ai_canonname;
  }

  unsigned char *address_bytes =
      reinterpret_cast<sockaddr_in6 *>(ai_.get()->ai_addr)->sin6_addr.s6_addr;

  for (size_t i = 0; i < sizeof(struct in6_addr); i++) {
    if (i > 1 && (i % 2) == 0) {
      expanded << std::setw(1) << ':';
    }
    expanded << std::setw(2) << std::setfill('0') << std::hex
             << (int)address_bytes[i];
  }

  return expanded.str();
}

bool IpAddress::is_ipv6() const { return ai_.get()->ai_family == AF_INET6; }

std::ostream &operator<<(std::ostream &os, const IpAddress &ip) {
  os << ip.get_expanded();
  return os;
}


etc/dibbler.sh
#!/bin/bash
#
# This script interprets DHCPv6 vendor options send from NTT for use in
# configuring the Softbank Hikari connection.
#

[ -f /etc/sysconfig/sofutobanku ] && . /etc/sysconfig/sofutobanku

LOGFILE=/var/log/dibbler-sofutobanku.log

action="$1"

adddate() {
    while IFS= read -r line; do
        printf "%s %s\n" "$(date +"%Y-%m-%d %T")" "$line"
    done
}

log() {
    echo $* | adddate >> $LOGFILE
}

parse_ntt_config() {
    local input="$@"
    for keyval in $input; do
        local key=${keyval%%=*}
        local val=${keyval#*=}
        case $key in
            vendor)
                if [[ $val != 210 ]]; then
                    log "Received vendor extension $val; exiting script..."
                    return
                fi
                ;;
            201)
                # MAC address
                mac_address="$val"
                ;;
        esac
    done

}

setup_local_address() {
    local ip_address=$1
    local old_ip_address="$(/sbin/ip -6 addr show dev $IFACE scope global)"
    /sbin/ip -6 addr replace $ip_address/64 dev $IFACE scope global
    local new_ip_address="$(/sbin/ip -6 addr show dev $IFACE scope global)"
    if [[ -z $old_ip_address ]]; then \
        log "Configured $new_ip_address/64 on interface $IFACE"
    elif [ $old_ip_address != $new_ip_address ]; then \
        log "Replaced $old_ip_address on interface $IFACE with new address $new_ip_address/64"
    fi
}

connect_tunnel() {
    local ip_address=$1
    local mac=$2
    /home/kenny/git/softubanku/build/sbradclient --username=$ip_address --password=$PASSWORD \
        --auth-server="$AUTH_SERVER" --shared-secret="$SHARED_SECRET" \
        --mac=$mac 2> >(adddate >> $LOGFILE) \
        | read local_ipv4 tunnel_endpoint
    log "Setting up tunnel to $tunnel_endpoint (local address is $local_ipv4)"
}

if [[ $action == add || $action == update ]]; then \
    parse_ntt_config ${SRV_OPTION17}

    if [[ -n $PREFIX1 ]]; then \
        local_address=${PREFIX1}1111:1111:1111:1111
        setup_local_address $local_address
    fi

    if [[ -r $AUTH_SERVER || -z $SHARED_SECRET || -z $PASSWORD ]]; then \
        log "/etc/sysconfig/sofutobanku must contain AUTH_SERVER, SHARED_SECRET, and PASSWORD; aborting..."
    elif [[ -n $local_address && -n $mac_address ]]; then \
        log "Did not receive IPv6 address or MAC address; aborting..."
    else \
        log "Trying to connect tunnel for $local_address"
        connect_tunnel $local_address $mac_address
    fi
fi

# vim: set et ts=4 sw=4 sts=4:
